cdc_iterator.go

```diff
@@ -0,0 +1,293 @@
+		}
+		if pos.Kind == common.PositionTypeSnapshot {
+			return mysql.Position{}, fmt.Errorf("invalid position type: %s", pos.Kind)
+		}
+
+		return pos.CdcPosition.Position, nil
+	}
+
+	masterPos, err := c.canal.GetMasterPos()
+	if err != nil {
+		return mysql.Position{}, fmt.Errorf("failed to get master position: %w", err)
+	}
+
+	return masterPos, nil
+}
+
+func (c *cdcIterator) runCanal(ctx context.Context, startPos mysql.Position) error {
+	// buffered to allow the goroutine to exit when the iterator is torn down
+	errChan := make(chan error, 1)
+	go func() {
+		handler := &cdcEventHandler{
+			ctx:  ctx,
+			data: c.data,
+		}
+		c.canal.SetEventHandler(handler)
+		errChan <- c.canal.RunFrom(startPos)
+	}()
+
+	select {
+	case <-ctx.Done():
+		c.canal.Close()
+		return fmt.Errorf("context cancelled from runCanal: %w", ctx.Err())
+	case err := <-errChan:
+		return fmt.Errorf("failed to run canal: %w", err)
+	}
+}
+
+func (c *cdcIterator) Ack(context.Context, sdk.Position) error {
+	c.acks.Done()
+	return nil
+}
+
+func (c *cdcIterator) Next(ctx context.Context) (sdk.Record, error) {
+	select {
+	case <-ctx.Done():
+		return sdk.Record{}, fmt.Errorf("context cancelled: %w", ctx.Err())
+	case data, ok := <-c.data:
+		if !ok { // closed
+			if err := c.canalTomb.Err(); err != nil {
+				return sdk.Record{}, fmt.Errorf("canal exited unexpectedly: %w", err)
+			}
+			if err := c.acks.Wait(ctx); err != nil {
+				return sdk.Record{}, fmt.Errorf("failed to wait for acks: %w", err)
+			}
+			return sdk.Record{}, fmt.Errorf("cdc iterator teared down")
+		}
+
+		rec, err := c.buildRecord(data)
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("failed to build record: %w", err)
+		}
+
+		c.acks.Add(1)
+		return rec, nil
+	}
+}
+
+func (c *cdcIterator) Teardown(context.Context) error {
+	if c.canalTomb != nil {
+		c.canalTomb.Kill(errors.New("tearing down snapshot iterator"))
+	}
+
+	return nil
+}
+
+func buildPayload(columns []schema.TableColumn, rows []any) sdk.StructuredData {
+	payload := sdk.StructuredData{}
+	for i, col := range columns {
+		val := rows[i]
+		if s, ok := val.(string); ok {
+			// I don't know why exactly, but "github.com/go-mysql-org/go-mysql/canal"
+			// returns a string for timestamp columns without timezone. This is a hack
+			// to format the string back into an UTC string.
+			// TODO: investigate this further. Is this a bug in canal?
+			val = tryParseCanalStrDate(s)
+		}
+
+		payload[col.Name] = common.FormatValue(val)
+	}
+	return payload
+}
+
+func tryParseCanalStrDate(s string) string {
+	parsed, err := time.Parse(time.DateTime, s)
+	if err != nil {
+		return s
+	}
+
+	valCopyInUTC := time.Date(
+		parsed.Year(), parsed.Month(), parsed.Day(),
+		parsed.Hour(), parsed.Minute(), parsed.Second(),
+		parsed.Nanosecond(),
+		time.Now().Location(),
+	).UTC()
+
+	return valCopyInUTC.Format(time.RFC3339)
+}
+
+func (c *cdcIterator) buildRecord(e *canal.RowsEvent) (sdk.Record, error) {
+	pos, err := c.canal.GetMasterPos()
```

Comment:
The position handling here feels sketchy. Are we sure this is correct? How do we know that the name in this position corresponds to the one that `e.Header.LogPos` points to? 🤔

Reply:
So, the short answer is that I haven't found the perfect way to maintain both the LogPos and the binary log name in sync.
The current method works as long as all `OnRow()` events that happen **before** the binary log name transition are processed, which there's no guarantee.

What I agree is that actually it is very ineficient to call `GetMasterPos()` for each OnRow event, that is basically polling.

Now, the best and most complete solution that I can think of to keep the log pos and the binary log name in sync would be to save every binary log name transition in a buffer with a timestamp. For every row event that I receive, I search for the corresponding binary log name instead of querying. I would keep it in sync with the other event that canal gives me, `OnRotate`.

You can see a list of events here:
https://github.com/go-mysql-org/go-mysql/blob/master/canal/handler.go

But I might be overcomplicating this.

Reply:
Oh, now I see, we should definitely use `OnRotate` to track the current binlog name.

Reply:
Btw, no need to store the timestamp, we just need to keep track of the last log name. Events should arrive in the correct order, so after we receive `OnRotate` all future events should belong to that log.

We also need to figure out how to consistently get the initial log name when first starting the connector, I feel like that's also a bit racy if we get unlucky.

Reply:
I don't see it documented anywhere, but reading `mysql.canal` sourcecode I can see that [canal.GetMasterPos()](https://github.com/go-mysql-org/go-mysql/blob/ee48e789c8edea845522cdab60303e63f61f45be/canal/sync.go#L318-L333) calls to Execute, which [creates a connection to mysql if not set](https://github.com/go-mysql-org/go-mysql/blob/ee48e789c8edea845522cdab60303e63f61f45be/canal/canal.go#L524-L529).

So, I guess that if I get the name from the master position before starting the canal we should be good to go right?

Reply:
But what if we get binlog file `01`, then the log rotates to `02` and then we start the canal? 🙈

Reply:
Aah yeah didn't consider that small window.

What if instead we:

1. start the canal first with an empty name in a goroutine. If in the `OnRow` event the name is empty, we wait for the name
2. get the master position in the parent goroutine and send it to the started goroutine
3. if in between 1 and 2 there was a log rotation, we ignore the sent name
4. otherwise, we set the binlog file to the received name.
5. send all OnRow events blocked in between.

Reply:
I took a step back and investigated [github.com/go-mysql-org/go-mysql-elasticsearch](https://github.com/go-mysql-org/go-mysql-elasticsearch) and played around with the lib and some tests. It looks like we don't need this call at all.

If we want to start listening to the binlog from a specific position, we already know the full position including the log name. That's relevant for when restarting a pipeline that was already running. So in that case we need to pass that log name to the event handler.

However, if we are starting a new pipeline where we don't have any stored position yet, there we can simply pass an empty position to `RunFrom` and we will receive [fake](https://github.com/mysql/mysql-server/blob/8e797a5d6eb3a87f16498edcb7261a75897babae/sql/rpl_binlog_sender.h#L235) rotate events until we reach the current binlog file.

To sum up - just take the start position from the config and leave it empty if it's not there. Everything else should be handled through the `OnRotate` event.

Reply:
But if we start from an empty position that means that we will have to handle all recorded events in mysql up to the latest position. I believe we do rather want all events that happen just after starting the snapshot transaction, right? Even with the future "snapshot": "never" parameter, we still would want to start from the latest position, at least with the default behaviour.

> But what if we get binlog file 01, then the log rotates to 02 and then we start the canal? 🙈 In other words, we missed the OnRotate event that happened between getting the master position and starting the **canal.**

At commit [d424177](https://github.com/conduitio-labs/conduit-connector-mysql/pull/22/commits/d4241777014dc68c74d3de546bba41611324b6d6) I added a test that should disprove this situation. So I believe we are actually good with what we have right now, or I'm still not understanding the root issue.